### PR TITLE
wordy: add to track

### PR DESCRIPTION
**Pull request by stkent**
Canonical data: https://github.com/exercism/x-common/blob/master/exercises/wordy/canonical-data.json

### CODE REVIEW

**Comment by jtigger** [comment]
BTW, as a reviewer, I really dig the reference to the canonical data.  When reviewing, I grab the description, that data and hit the test suite, first off.  Having the link in the PR is a friendly optimization.

Thanks, man!

File: exercises/wordy/build.gradle [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  mavenCentral()
+}
+
+dependencies {
+  testCompile "junit:junit:4.12"
+}
+test {
+  testLogging {
+    exceptionFormat = 'full'
+    events = ["passed", "failed", "skipped"]
```
Comment (stkent) [inline exercises/wordy/build.gradle]: I added `showStandardStreams = true` here while I was testing my implementation; could be something we consider as standard in the future to help folks?

File: exercises/wordy/src/example/java/WordProblemSolver.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.regex.Pattern;
+
+final class WordProblemSolver {
+
+    private static final String INTEGER_REGEX_STRING = "(-?\\d+)";
+
+    private static final String BINARY_OPERATION_REGEX_STRING
+        = "(plus|minus|multiplied by|divided by|raised to the power)";
+
+    private static final String VALID_QUESTION_REGEX_STRING =
```
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Splitting into components allows for the entire problem to be validated, as well as the individual pieces of information to be extracted for the actual calculations. For example, not checking the overall structure could allow a question like

> What is 3 nonsense words lorem ipsum plus 4?

to be considered valid. Still feels quite awkward though.

I initially tried writing a single regex string that would grab every piece of information required at once while simultaneously validating the entire problem structure, but then I ran into difficulties with [backtracking](http://www.regular-expressions.info/captureall.html).

File: exercises/wordy/src/example/java/WordProblemSolver.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            "^What is " + INTEGER_REGEX_STRING +
+            "( " + BINARY_OPERATION_REGEX_STRING + " " + INTEGER_REGEX_STRING + ")+" +
+            "\\?$";
+
+    int solve(final String wordProblem) {
+        if (!wordProblem.matches(VALID_QUESTION_REGEX_STRING)) {
+            throw new IllegalArgumentException("I'm sorry, I don't understand the question!");
```
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: 😢

File: exercises/wordy/src/example/java/WordProblemSolver.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        final Matcher initialValueMatcher = Pattern.compile(INTEGER_REGEX_STRING).matcher(wordProblem);
+
+        initialValueMatcher.find();
```
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Guaranteed to find _something_, because we know the question has the appropriate structure.

File: exercises/wordy/src/example/java/WordProblemSolver.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        int result = Integer.parseInt(initialValueMatcher.group());
+
+        final Matcher operationAndValueMatcher
+                = Pattern.compile(BINARY_OPERATION_REGEX_STRING + " " + INTEGER_REGEX_STRING).matcher(wordProblem);
```
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Finds strings like "divided by 5" and captures the groups "divided by" and "5".

File: exercises/wordy/src/example/java/WordProblemSolver.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        switch (operationDescription) {
+            case "plus":
+                return firstValue + secondValue;
+            case "minus":
+                return firstValue - secondValue;
+            case "multiplied by":
+                return firstValue * secondValue;
+            case "divided by":
+                return firstValue / secondValue;
+            case "raised to the power":
```
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Exponentiation was noted as extra credit in the common tests due to needing a different sentence structure, but it is also quite common to say "2 raised to the power 5", so I threw it in as a regular test in that format instead. Let me know if this deviation is dangerous; if so, I'll probably pull powers entirely.

File: exercises/wordy/src/test/java/WordProblemSolverTest.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        assertEquals(32, new WordProblemSolver().solve("What is 2 raised to the power 5?"));
+    }
+
+    @Ignore
+    @Test
+    public void testInvalidQuestionFormat() {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("I'm sorry, I don't understand the question!");
+
+        // See https://en.wikipedia.org/wiki/President_of_the_United_States if you really need to know!
```
Comment (stkent) [inline exercises/wordy/src/test/java/WordProblemSolverTest.java]: Distracting, or fun?

**Comment by stkent** [comment]
Updated with implementation, and ready for review!

**Comment by jtigger** [comment]
I've been out of commission with a cold for a number of days.  All I had the energy for was tweeting. ;)  After a nap, here, I'll give ya a proper review, @stkent; sorry for the delay.

**Comment by stkent** [comment]
No rush; feel better first!

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 2 above)
Comment (jtigger) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: insert obligatory line about regexes... like if you've got a problem then try to solve it with a regex, now you got two problems. :)

It's great mental push-ups, but likely too dense to be maintainable.

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 3 above)
Comment (jtigger) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Why the long face?

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 4 above)
Comment (jtigger) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Such a great example of non-defensive programming (which increases readability).

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 5 above)
Comment (jtigger) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: You have *fun* with this one! :)  So delighted you're adding this exercise to this track, @stkent !

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 6 above)
Comment (jtigger) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: I believe we should have a complete solution in the example code.  So, keep it, please.

File: exercises/wordy/src/test/java/WordProblemSolverTest.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    @Ignore
+    @Test
+    public void testMultipleDivisions() {
+        assertEquals(2, new WordProblemSolver().solve("What is -12 divided by 2 divided by -3?"));
+    }
+
+    @Ignore
+    @Test
+    public void testExponentiation() {
+        assertEquals(32, new WordProblemSolver().solve("What is 2 raised to the power 5?"));
```
Comment (jtigger) [inline exercises/wordy/src/test/java/WordProblemSolverTest.java]: What do you think of splitting this out into a separate test class whose name clearly indicates it's bonus tests that can remain ignored?

Also, I'm wondering if we'd like to expand a little on the canonical data set and include more cases for this.  This is inbounds, AFAIK... it's okay to provide additional cases as long as it is sufficiently additive to the experience.  I think the expectation is that we contribute back to x-common.

In general, `wordy` seems to need some love.  I've already submitted a PR to make the x-common descriptive bits... want to address some the weaknesses in the canonical data as well.

File: exercises/wordy/src/test/java/WordProblemSolverTest.java (hunk 7 above)
Comment (jtigger) [inline exercises/wordy/src/test/java/WordProblemSolverTest.java]: FUN!

**Comment by jtigger** [comment]
There was a question about maybe adding a few bits in `HINT.md`... @stkent can you please indicate whether you want this in the PR or no?  Ready to merge.

File: exercises/wordy/build.gradle (hunk 1 above)
Comment (stkent) [inline exercises/wordy/build.gradle]: Sounds good, will keep it in mind!

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 2 above)
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Agreed!

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 3 above)
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Our word problem solver is only so smart!

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 5 above)
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: Haha yeah, this was a good exercise to build!

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 6 above)
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: I'm not sure I clearly described what I had hoped to here. Let me try again:

In the canonical tests, exponentiation is included as extra credit based on the fact that supporting a sentence structure like "What is 52 cubed?" would require additional sentence structure matching. However, since I noted that "What is 52 raised to the power 3?" fit the existing structure, I wrote a test for that case which _doesn't_ require new matching code. This felt opportunistic but is a deviation from the canonical data and its intent.

@jtigger does that perturb you more? I think it should.

File: exercises/wordy/src/test/java/WordProblemSolverTest.java (hunk 7 above)
Comment (stkent) [inline exercises/wordy/src/test/java/WordProblemSolverTest.java]: Hooray! :D

**Comment by stkent** [comment]
@jtigger replied! One comment requires further discussion before merge. The `HINT.md` adds can be added separately after #142, I think.

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 3 above)
Comment (jtigger) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: ah! yes. :)

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 6 above)
Comment (jtigger) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: I'll blame the cold. 😬 I now understand your original question.

It's not obvious to me...

- it's a bonus iteration, so this would indicate to me a requirement that likely causes the practitioner to rethink their whole solution.
  - In general, I wish we had something like this for all exercises to both encourage exploration and increase fluency in the skill of deferring design choices that can be deferred (i.e. make the solution as concrete as feasible and generalize as requirements arrive).
  - even better is a pair of conflicting bonuses
- the practitioner is being asked to write a parser, if the requirement encourages them to "mature" that part of the code, seems on target.
- however, it's kindof a one-off... while that is a nice simulation of real-world software, it's a bit frustrating to the Exercism practitioner.

In the sum, I personally come down on the side of supporting the additional sentence structure.

_That said_, it is truly up to you: if you dig writing the more general solution, go for it.  If you either disagree with the sum or you agree but want to get this off your plate,  we can take this for now and amend the exercise later (if someone wants to).  There's no additional cost to deferring.

**Comment by jtigger** [comment]
@stkent back to ya.

File: exercises/wordy/src/example/java/WordProblemSolver.java (hunk 6 above)
Comment (stkent) [inline exercises/wordy/src/example/java/WordProblemSolver.java]: @jtigger thanks for the thoughts! Agreed that this one is far from clear-cut.

I'm in the "want to get this off my 🍽" camp at this point in time, so I've simplified the implementation and replaced the old exponentiation test with a new one. PR should be good for another look!

**Comment by jtigger** [comment]
So, I attempted to use GitHub's new "resolve conflict" feature and apparently failed. :)  Looking into it.

**Comment by jtigger** [comment]
Thank you, @stkent !!!
